tools/nl/embeddings/palm_alternatives.py

```diff
@@ -0,0 +1,276 @@
+  return similar_inds
+
+
+def update_local_sheets_file(ctx, sheets_url: str, worksheet_name: str,
+                             local_sheets_filepath: str) -> None:
+  """Download latest from Google sheets file and write to the local csv file."""
+  df = _get_sheets_data(ctx, sheets_url, worksheet_name)
+  df = df.fillna("")
+  print(f"Downloaded {len(df)} lines.")
+
+  print("Updating input CSV file")
+  df.to_csv(local_sheets_filepath, index=False)
+
+
+def load_svs(local_sheets_filepath: str) -> pd.DataFrame:
+  return pd.read_csv(local_sheets_filepath)
+
+
+def build_embeddings(ctx, texts):
+  print("Building embeddings.")
+  embeddings = ctx.model.encode(texts, show_progress_bar=True)
+  embeddings = pd.DataFrame(embeddings)
+  return torch.from_numpy(embeddings.to_numpy()).to(torch.float)
+
+
+def add_palm_alternatives(ctx, df_svs: pd.DataFrame) -> None:
+  """Generate alternative descriptions using the PaLM API and add as a column to df_svs.
+  
+  Args:
+    df_svs: DataFrame with the latest SVs, descriptions and other human curated alternatives.
+    palm_alternatives_filepath: the filepath to write the PaLM alternatives to.
+  """
+  (texts, dcids) = _get_texts_dcids(df_svs, SHEETS_DCID_COL,
+                                    SHEETS_ALTERNATIVES_COL)
+  embeddings = build_embeddings(ctx, texts)
+
+  # Start processing.
+  palm_alternatives = []
+  total_rows = len(df_svs)
+  for rid, row in df_svs.iterrows():
+    if rid % 10 == 0:
+      print(f"Processed: {rid}. Index = {rid}/{total_rows}")
+      time.sleep(API_TIMEOUT_SECONDS)
+
+    sv_dcid = row[SHEETS_DCID_COL]
+    sv_description = row[SHEETS_DESCRIPTION_COL]
+
+    # if the description is empty, use the first alternative.
+    if not sv_description:
+      sv_description = row[SHEETS_ALTERNATIVES_COL].split(";")[0]
+
+    # Do not generate alternatives for topics.
+    if 'dc/topic' in sv_dcid:
+      continue
+    print(f"***Processing sv_dcid:{sv_dcid}, sv_description:{sv_description}")
+
+    # Get Alternative.
+    valid_alts = []
+    for _ in range(MAX_API_RETRIES):
+      alts = palm_helper.palm_api_call(PALM_API_URL, sv_description,
+                                       TEMPERATURE, API_TIMEOUT_SECONDS)
+
+      for a in alts:
+        # Check if we can find any valid alternates.
+        matched_inds = _sv_match_alternative(ctx,
+                                             sv_description,
+                                             embeddings,
+                                             top_k=3)
+
+        top_sv_id = dcids[matched_inds[0]]
+        if top_sv_id == sv_dcid:
+          valid_alts.append(a)
+        else:
+          # Some print logs in case of no matches.
+          print(f"  -- Not valid alternative: {a}")
+          print(
+              f"  -- Top Alternative was: {top_sv_id}. Expected SV: {sv_dcid}")
+
+      if len(valid_alts) >= MAX_ALTERNATIVES_LIMIT:
+        break
```

Comment:
nit: worth also breaking in the inner for loop to keep valid_alts at MAX_ALTERNATIVES_LIMIT?

Reply:
Done